radix-engine/src/transaction/transaction_executor.rs

```diff
@@ -116,28 +118,40 @@ where
         #[cfg(not(feature = "alloc"))]
         let now = std::time::Instant::now();
 
-        // The transaction should already have been verified against the network
-        // TODO: Replace this with a rejection, for tests
-        assert_eq!(
-            transaction.transaction_network_id(),
-            params.network_definition.id,
-            "Transaction network id didn't match configuration"
-        );
-
         let transaction_hash = transaction.transaction_hash();
-        let transaction_network = &params.network_definition;
+        let transaction_network_id = transaction.transaction_network_id();
         let signer_public_keys = transaction.signer_public_keys().to_vec();
         let instructions = transaction.instructions().to_vec();
         #[cfg(not(feature = "alloc"))]
         if params.trace {
             println!("{:-^80}", "Transaction Metadata");
             println!("Transaction hash: {}", transaction_hash);
-            println!("Transaction network: {:?}", transaction_network);
+            println!("Transaction network id: {}", transaction_network_id);
             println!("Transaction signers: {:?}", signer_public_keys);
 
             println!("{:-^80}", "Engine Execution Log");
         }
 
+        let transaction_contents = TransactionContents {
+            network_id: transaction_network_id,
+            instructions: instructions,
+        };
+
+        // 0. Precondition checks
+        if params.network_definition.id != transaction.transaction_network_id() {
+            return Self::compose_receipt(
+                params,
+                transaction_contents,
+                None,
+                TransactionResult::Reject(RejectResult {
+                    error: RejectionError::MismatchedNetwork {
+                        transaction_network_id: transaction.transaction_network_id(),
+                        expected_network_id: params.network_definition.id,
+                    },
+                }),
+            );
+        }
```

Comment:
Don't think we need to validate the transaction network id here. It's part of the static header validation process, same as other header fields. TransactionExecutor only issue stateful execution-based rejections.

Given network ID is no longer used or exposed to Scrypto, network definition is no longer required by TransactionExecutor.

Reply:
Hmmm fair, but the receipt needs to somehow know the network configuration in order to format itself, and the header only contains the network id at present.

So the question is:
1. Where does that configuration come from?
2. How do we sanity-check we're using the right one, to limit possibility for confusion and provide defence-in-depth?

Where does it come from = either the (a) Executable Transaction or the (b) ExecutionParams.

If (a), we'd need to clone it inside ValidatedTransaction and TestTransaction.

If (b), the only real way to achieve (2) is to do as above -- otherwise someone could run the transaction through and get a receipt for a different network to the header; which would be weird.

I went for (b) because the network configuration felt like it belonged as a parameter, configuring the engine. But you're saying possibly that's not the direction of vision/travel? Would you prefer option (a) in that case?

Reply:
> The receipt needs to somehow know the network configuration in order to format itself

It can take a network id parameter. Formatting of addresses are broken ATM (hardcoded network id in `ResourceAddress`/`PackageAddress`).

> How do we sanity-check we're using the right one, to limit possibility for confusion and provide defence-in-depth?

It's all part of the `TransactionValidator` (and `ValidationConfig`).

Reply:
For anyone else tracking this, Yulong and I caught up:
* Formatting is currently broken (see eg) format! of PackageAddress which assumes LocalSim
* For now, let's break formatting of the receipt further, and have a separate task

Therefore:
* Remove Network Definition from the execution params and executor for this PR

We'll have future conversations about:
* How to format stuff correctly
* Validation - particularly of Network Id and Epoch; and whether that should be in the Executor or Validator (in particular, so dependency tracking works right for a "clever" mempool)